Add onboarding/plugins REST endpoint

moon0326 opened the pull request:
### Changes proposed in this Pull Request:

This PR adds new REST endpoints for onboarding plugin installation page.

We already have async install and status check endpoints in [Plugins.php](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/API/Plugins.php), but we wanted to have dedicated endpoints for the onboarding experience with real-time error check.

1. Added the following endpoints.

POST /wp-json/wc-admin/onboarding/plugins/install-async
GET /wp-json/wc-admin/onboarding/plugins/scheduled-installs/:job-id

You can find the API spec [here](https://github.com/woocommerce/team-ghidorah/issues/212)

2. Updated [PluginsHelper::install](https://github.com/woocommerce/woocommerce/pull/38174/files#diff-38f216066764e2afb70406832279e29873dc34008334384bfd79011a95188bbbR155) to accept an instance of [PluginsInstallLogger](https://github.com/woocommerce/woocommerce/pull/38174/files#diff-e9b97940cda314d3b826232f50920d76b0bbb72ac9df1970e446b7de61956c02). At the moment, we only use [AsyncPluginsInstallLogger](https://github.com/woocommerce/woocommerce/pull/38174/files#diff-a0e49d943df2b73567b8052b507c6a578a06421a9e940c9e65ec207288e97991)

Closes https://github.com/woocommerce/team-ghidorah/issues/212

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

1. Make a request to POST `/wp-json/wc-admin/onboarding/plugins/install-async` with the following JSON request

```
{
    "plugins": ["test"]
}
```

2. The response should contain `job_id` field. Copy it.
3. Make a request to GET `/wp-json/wc-admin/onboarding/plugins/scheduled-installs/:job_id`. It should return the following response.

```
{
    "job_id": ":job_id",
    "status": "complete",
    "plugins": {
        "test": {
            "status": "failed",
            "errors": [
                "The requested plugin `test` could not be installed. Plugin API call failed."
            ],
            "install_duration": 0
        }
    }
}
```

4. Repeat steps 1 ~ 3, but with a valid plug-in name this time.

<!-- End testing instructions -->

Commit:
Remove unused $job_id argument

Commit:
Support WP_Upgrader argument -- helps for testing

Commit:
Minor refactor to support a custom logger in install_plugins

   * Support a custom logger in install_plugins
   * Support swapping out WP_Upgrader impl.
   * Added woocommerce_plugins_install_async_callback callback to call install_plugins with a custom logger from action scheduler

Commit:
Add a new REST endpoint for onboarding plugins

* POST /wc-admin/onboarding/plugins/install-async -- queues plugin installs to action scheudler
* GET /wc-admin/onboarding/plugins/scheduled-installs/:job_id -- returns current status of given job id

Commit:
Remove WP_Upgrader support -- not using it

Commit:
Add changelog

Commit:
Fix phpcs errors

Commit:
Call complete

Commit:
Update test to wait 1 sec before checking the action scheduler

Commit:
Remove test that requires waiting an async action

Commit:
Add test for 404 status

chihsuan commented on plugins/woocommerce/tests/php/src/Admin/API/OnboardingPluginsTest.php:
```diff
@@ -0,0 +1,111 @@
+use WP_REST_Request;
+
+/**
+ * MarketingCampaigns API controller test.
```
```suggestion
 * OnboardingPlugins API controller test.
```

chihsuan commented on plugins/woocommerce/tests/php/src/Admin/API/OnboardingPluginsTest.php:
```diff
@@ -0,0 +1,111 @@
+/**
+ * MarketingCampaigns API controller test.
+ *
+ * @class MarketingCampaignsTest.
```
```suggestion
 * @class OnboardingPluginsTest.
```

chihsuan commented on plugins/woocommerce/tests/php/src/Admin/API/OnboardingPluginsTest.php:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	/**
+	 * Request to schedulled-installs endpoint.
```
```suggestion
	 * Request to scheduled-installs endpoint.
```

chihsuan commented on plugins/woocommerce/src/Admin/PluginsInstallLoggers/PluginsInstallLogger.php:
```diff
@@ -0,0 +1,43 @@
+	public function install_requested( string $plugin_name );
+
+	/**
+	 * Caleld when a plugin installed sucessfully.
```
```suggestion
	 * Called when a plugin installed successfully.
```

chihsuan commented on plugins/woocommerce/src/Admin/PluginsInstallLoggers/PluginsInstallLogger.php:
```diff
@@ -0,0 +1,43 @@
+	public function installed( string $plugin_name, int $duration);
+
+	/**
+	 * Called when an error ocured while installing a plugin.
```
```suggestion
	 * Called when an error occurred while installing a plugin.
```

chihsuan commented on plugins/woocommerce/src/Admin/PluginsInstallLoggers/AsynPluginsInstallLogger.php:
```diff
@@ -0,0 +1,113 @@
+/**
+ * A logger to log plugin installation progress in real time to an option.
+ */
+class AsynPluginsInstallLogger implements PluginsInstallLogger {
```
Nit: Perhaps, we should use `Async`?

```suggestion
class AsyncPluginsInstallLogger implements PluginsInstallLogger {
```

chihsuan reviewed (approved):
Nice work! LGTM and tested well. 👍 

Left some small comments (typos) pre-approving!

chihsuan reviewed (dismissed):
Nice work! LGTM and tested well. 👍 

Left some small comments (typos) pre-approving!

chihsuan commented:
Hey @moon0326 This is unrelated to this PR, but I'm curious about when do we activate plugins?

moon0326 commented:
Thank you for the review @chihsuan 👍 

> Hey @moon0326 This is unrelated to this PR, but I'm curious about when do we activate plugins?

That's a good question.

If installation completes within 30 seconds, we can activate the plugins from the client. Otherwise, async endpoints should take care of that. I'll make a change for that use case.

ilyasfoo commented on plugins/woocommerce/src/Admin/PluginsHelper.php:
```diff
@@ -285,7 +319,7 @@ public static function schedule_install_plugins( $plugins ) {
 
 		$job_id = uniqid();
-		WC()->queue()->schedule_single( time() + 5, 'woocommerce_plugins_install_callback', array( $plugins, $job_id ) );
+		WC()->queue()->schedule_single( time() + 5, 'woocommerce_plugins_install_callback', array( $plugins ) );
```
Good catch! I guess this wasn't used before?

moon0326 replied:
Yup! $job_id was never used in the callback function.

ilyasfoo reviewed (commented):
Awesome work, @moon0326! So far testing well per instruction, however, I'm noticing the `install-async` endpoint is taking a significantly different amount of time to respond when I add a few plugins to install compared to only one. I think it might still be synchronous somewhere.

moon0326 commented:
> Awesome work, @moon0326! So far testing well per instruction, however, I'm noticing the `install-async` endpoint is taking a significantly different amount of time to respond when I add a few plugins to install compared to only one. I think it might still be synchronous somewhere.

That's interesting. I'll look into it.

Commit:
Add status field

Commit:
Remove woocommerce_plugins_install_async_callback action -- no longer used

Commit:
Call run-in-background endpoint with blocking=false to simulate a background process

moon0326 commented:
@chihsuan @ilyasfoo 

I've updated the code with a different approach.

1. Removed ActionScheduler.
2. Add [run_in_background](https://github.com/woocommerce/woocommerce/pull/38174/files#diff-ccca16b15b2dfeb741b98e5d2695dbf1d789de8a44fb869e426929f62f681863R85) endpoint. It is called in `install-async- endpoint with [blocking = false](https://github.com/woocommerce/woocommerce/pull/38174/files#diff-ccca16b15b2dfeb741b98e5d2695dbf1d789de8a44fb869e426929f62f681863R131) to simulate a background process[

This worked pretty well for me.

Commit:
Update plugins/woocommerce/src/Admin/PluginsInstallLoggers/AsynPluginsInstallLogger.php

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

Commit:
Update plugins/woocommerce/src/Admin/PluginsInstallLoggers/PluginsInstallLogger.php

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

Commit:
Update plugins/woocommerce/tests/php/src/Admin/API/OnboardingPluginsTest.php

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

Commit:
Update plugins/woocommerce/tests/php/src/Admin/API/OnboardingPluginsTest.php

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

Commit:
Update plugins/woocommerce/tests/php/src/Admin/API/OnboardingPluginsTest.php

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

Commit:
Update plugins/woocommerce/src/Admin/PluginsInstallLoggers/PluginsInstallLogger.php

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

Commit:
Fix incorrect logger name

Commit:
Add max_execution_time setting

Commit:
Remove test code

Commit:
Skip failing test -- hard to test as the function uses site url and test env does not actaully have a running WP

Commit:
code format

Commit:
Revert background process changes

Commit:
Add install and activate endpoint

moon0326 commented:
@chihsuan @ilyasfoo 

Please disregard my last comment. I think it's best to go with a hybrid approach to support broad # of use cases.

```
Given X number of plugins
Set 30 seconds timer
Request install-and-activate endpoint for each plugin
If we don't install everything within 30 seconds
Queue the remaining plugins via `install-async` endpoint
If there is ongoing installation, queue that as well.

```

I've updated the code to fix `install-async` endpoint to make it async. I've removed [ActionScheduler::runner()->run();](https://github.com/woocommerce/woocommerce/commit/baa0e608794e38c843917e46b44d5846928e8a50#diff-ccca16b15b2dfeb741b98e5d2695dbf1d789de8a44fb869e426929f62f681863R101) that blocked the execution.

I've also added [install-and-activate](https://github.com/woocommerce/woocommerce/pull/38174/files#diff-ccca16b15b2dfeb741b98e5d2695dbf1d789de8a44fb869e426929f62f681863R73) endpoint to install and activate in a single request.

I'll update [[Extensions] - Create "Get a boost with our free features" page#211](https://github.com/woocommerce/team-ghidorah/issues/211) issue about hybrid approach

ilyasfoo commented on plugins/woocommerce/src/Admin/API/OnboardingPlugins.php:
```diff
@@ -0,0 +1,312 @@
+					'properties' => array(
+						'errors'     => array(
+							'type'  => 'array',
+							'items' => array( 'type' => 'string' ),
```
I'm wondering if we need to specify which plugin the error is for, do we need to highlight the plugins affected?

moon0326 replied:
Thank you for catching it. Fixed in https://github.com/woocommerce/woocommerce/pull/38174/commits/b4f66e27fbed915b78238b25657fec7b86ea47e3

ilyasfoo commented on plugins/woocommerce/src/Admin/PluginsInstallLoggers/AsynPluginsInstallLogger.php:
```diff
@@ -0,0 +1,134 @@
+/**
+ * A logger to log plugin installation progress in real time to an option.
+ */
+class AsynPluginsInstallLogger implements PluginsInstallLogger {
```
It seems the change got replaced with a rebase?

```suggestion
class AsyncPluginsInstallLogger implements PluginsInstallLogger {
```

ilyasfoo commented on plugins/woocommerce/src/Admin/API/OnboardingPlugins.php:
```diff
@@ -0,0 +1,312 @@
+	 * @return WP_Error|boolean
+	 */
+	public function can_install_plugins() {
+		return true;
```
I think this is a testing line? 😄 
```suggestion
```

moon0326 replied:
Thank you for cathing it 🙇

ilyasfoo commented on plugins/woocommerce/src/Admin/API/OnboardingPlugins.php:
```diff
@@ -0,0 +1,312 @@
+				array(
+					'methods'             => 'POST',
+					'callback'            => array( $this, 'install_and_activate' ),
+					'permission_callback' => array( $this, 'can_install_plugins' ),
```
I think we'll need to also check `current_user_can( 'activate_plugins' )` for this particular endpoint

moon0326 replied:
Good suggestion! Updated in https://github.com/woocommerce/woocommerce/pull/38174/commits/1355fe3c963e54ce715091bed2457d2b45fe323b

ilyasfoo commented on plugins/woocommerce/tests/php/src/Admin/API/OnboardingPluginsTest.php:
```diff
@@ -0,0 +1,114 @@
+		);
+		wp_set_current_user( $this->user );
+	}
+
```
Since `install_and_activate` and `install_async` endpoints pose a greater risk if the permission check fails, should we add a test to make sure all these endpoints are protected by the relevant permission checks?

moon0326 replied:
Good idea!

Updated in https://github.com/woocommerce/woocommerce/pull/38174/commits/765c2daef986bc63f6ea0ef4c7c17b01acbd9092

ilyasfoo reviewed (changes requested):
Excellent work, @moon0326! The async installation works well, I was able to get a quick response on the API calls. Some minor comments:

Commit:
Update plugins/woocommerce/src/Admin/PluginsInstallLoggers/AsynPluginsInstallLogger.php

Co-authored-by: Ilyas Foo <foo.ilyas@gmail.com>

Commit:
Update plugins/woocommerce/src/Admin/API/OnboardingPlugins.php

Co-authored-by: Ilyas Foo <foo.ilyas@gmail.com>

Commit:
Require activate_plugins permission

Commit:
Add test to check for permissions

Commit:
Update install and activate schema

ilyasfoo reviewed (changes requested):
Awesome, thanks for addressing @moon0326!

I'm getting this error in my action scheduler:

<img width="997" alt="image" src="https://github.com/woocommerce/woocommerce/assets/3747241/3f3441e6-73b2-4c82-877e-c86f4672be60">


```
Unexpected shutdown: PHP Fatal error
Uncaught Error: Class 'Automattic\WooCommerce\Admin\PluginsInstallLoggers\AsynPluginsInstallLogger' not found in /var/www/html/wp-content/plugins/woocommerce/src/Admin/PluginsHelper.php:331

Stack trace:
#0 /var/www/html/wp-includes/class-wp-hook.php(308): Automattic\WooCommerce\Admin\PluginsHelper::install_plugins_async_callback(Array, '646449826503b')
#1 /var/www/html/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters('', Array)
#2 /var/www/html/wp-includes/plugin.php(565): WP_Hook->do_action(Array)
#3 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/actions/ActionScheduler_Action.php(43): do_action_ref_array('woocommerce_plu...', Array)
#4 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/abstracts/ActionScheduler_Abstract_QueueRunner.php(65): ActionScheduler_Action->execute()
#5 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_QueueRunner.php(166): ActionScheduler_Abstract_QueueRunner->process_a in /var/www/html/wp-content/plugins/woocommerce/src/Admin/PluginsHelper.php on line 331
```

Commit:
Fix: import AsyncPluginsInstallLogger correctly

moon0326 commented:
> Awesome, thanks for addressing @moon0326!
> 
> I'm getting this error in my action scheduler:
> 
> <img alt="image" width="997" src="https://user-images.githubusercontent.com/3747241/238825035-3f3441e6-73b2-4c82-877e-c86f4672be60.png">
> ```
> Unexpected shutdown: PHP Fatal error
> Uncaught Error: Class 'Automattic\WooCommerce\Admin\PluginsInstallLoggers\AsynPluginsInstallLogger' not found in /var/www/html/wp-content/plugins/woocommerce/src/Admin/PluginsHelper.php:331
> 
> Stack trace:
> #0 /var/www/html/wp-includes/class-wp-hook.php(308): Automattic\WooCommerce\Admin\PluginsHelper::install_plugins_async_callback(Array, '646449826503b')
> #1 /var/www/html/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters('', Array)
> #2 /var/www/html/wp-includes/plugin.php(565): WP_Hook->do_action(Array)
> #3 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/actions/ActionScheduler_Action.php(43): do_action_ref_array('woocommerce_plu...', Array)
> #4 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/abstracts/ActionScheduler_Abstract_QueueRunner.php(65): ActionScheduler_Action->execute()
> #5 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_QueueRunner.php(166): ActionScheduler_Abstract_QueueRunner->process_a in /var/www/html/wp-content/plugins/woocommerce/src/Admin/PluginsHelper.php on line 331
> ```

My bad. I forgot to import AsyncPluginsInstallLogger correctly. Fixed in https://github.com/woocommerce/woocommerce/pull/38174/commits/0d2221121f66820c9bbfbb267553353e2bb90720

Commit:
Fix: add missing comments

ilyasfoo reviewed (approved):
Nice, works well now! LGTM 🚢
